examples/ping.rs

```diff
@@ -0,0 +1,139 @@
+#[macro_use]
+extern crate log;
+extern crate env_logger;
+extern crate getopts;
+extern crate smoltcp;
+extern crate byteorder;
+
+mod utils;
+
+use std::str::{self, FromStr};
+use std::time::{Duration, Instant};
+use smoltcp::Error;
+use smoltcp::wire::{EthernetAddress, IpAddress, IpProtocol, Icmpv4Repr, Icmpv4Packet};
+use smoltcp::iface::{ArpCache, SliceArpCache, EthernetInterface};
+use smoltcp::socket::{AsSocket, SocketSet};
+use smoltcp::socket::{RawSocket, RawSocketBuffer, RawPacketBuffer};
+use std::collections::HashMap;
+use byteorder::{ByteOrder, NetworkEndian};
+
+const PING_INTERVAL_S: u64 = 1;
+const PING_TIMEOUT_S: u64 = 5;
+const PINGS_TO_SEND: usize = 4;
+
+fn main() {
+    utils::setup_logging();
+
+    let (device, args) = utils::setup_device(&["ADDRESS"]);
+    let remote_addr = IpAddress::from_str(&args[0]).expect("invalid address format");
+    let local_addr = IpAddress::v4(192, 168, 69, 2);
+
+    let raw_rx_buffer = RawSocketBuffer::new(vec![RawPacketBuffer::new(vec![0; 64])]);
+    let raw_tx_buffer = RawSocketBuffer::new(vec![RawPacketBuffer::new(vec![0; 128])]);
+    let raw_socket = RawSocket::new(raw_rx_buffer, raw_tx_buffer, local_addr, IpProtocol::Icmp);
+
+    let hardware_addr = EthernetAddress([0x02, 0x00, 0x00, 0x00, 0x00, 0x02]);
+    let arp_cache = SliceArpCache::new(vec![Default::default(); 8]);
+    let mut iface = EthernetInterface::new(
+        Box::new(device),
+        Box::new(arp_cache) as Box<ArpCache>,
+        hardware_addr,
+        [local_addr],
+    );
+
+    let mut sockets = SocketSet::new(vec![]);
+    let raw_handle = sockets.add(raw_socket);
+
+    let mut send_next = Duration::default();
+    let mut seq = 0;
+    let mut received = 0;
+    let mut echo_payload = [0xffu8; 40];
+    let mut waiting_queue = HashMap::new();
+
+    let startup_time = Instant::now();
+
+    loop {
+        {
```

Comment:
It's necessary because both `sockets.get_mut` and `iface.poll` take a mutable reference to `sockets`. It's the same as in the `client.rs` example.